UseCases/Sources/ManualContextPropagation/main.swift

```diff
@@ -0,0 +1,130 @@
+    func inject(from context: Context, into: inout InjectInto) {
+        self.inject(context, &into)
+    }
+}
+
+typealias HTTPHeaders = [(String, String)]
+
+// MARK: - Fake HTTP Server
+
+struct FakeHTTPRequest {
+    let path: String
+    var headers: HTTPHeaders
+}
+
+struct FakeHTTPResponse {}
+
+typealias HTTPHeadersIntrumentationMiddleware = InstrumentationMiddleware<HTTPHeaders, HTTPHeaders>
+
+struct FakeHTTPServer {
+    typealias Handler = (Context, FakeHTTPRequest, FakeHTTPClient) -> FakeHTTPResponse
+
+    private let instrumentationMiddlewares: [InstrumentationMiddleware<HTTPHeaders, HTTPHeaders>]
+    private let catchAllHandler: Handler
+    private let client: FakeHTTPClient
+
+    init<M: InstrumentationMiddlewareProtocol>(
+        instrumentationMiddlewares: [M],
+        catchAllHandler: @escaping Handler
+    ) where M.InjectInto == HTTPHeaders, M.ExtractFrom == HTTPHeaders {
+        self.instrumentationMiddlewares = instrumentationMiddlewares.map {
+            InstrumentationMiddleware(extract: $0.extract, inject: $0.inject)
+        }
+        self.catchAllHandler = catchAllHandler
+        self.client = FakeHTTPClient(instrumentationMiddlewares: instrumentationMiddlewares)
+    }
+
+    func receive(_ request: FakeHTTPRequest) {
+        var context = Context()
+        print("\(String(describing: Self.self)): Extracting context values from request headers into context")
+        instrumentationMiddlewares.forEach { $0.extract(from: request.headers, into: &context) }
+        _ = catchAllHandler(context, request, client)
+    }
+}
+
+// MARK: - Fake HTTP Client
+
+struct FakeHTTPClient {
+    private let instrumentationMiddlewares: [InstrumentationMiddleware<HTTPHeaders, HTTPHeaders>]
+
+    init<M: InstrumentationMiddlewareProtocol>(
+        instrumentationMiddlewares: [M]
+    ) where M.InjectInto == HTTPHeaders, M.ExtractFrom == HTTPHeaders {
+        self.instrumentationMiddlewares = instrumentationMiddlewares.map {
+            InstrumentationMiddleware(extract: $0.extract, inject: $0.inject)
+        }
+    }
+
+    func performRequest(_ context: Context, request: FakeHTTPRequest) {
+        var request = request
+        print("\(String(describing: Self.self)): Injecting context values into request headers")
+        instrumentationMiddlewares.forEach { $0.inject(from: context, into: &request.headers) }
+        print(request)
+    }
+}
+
+// MARK: - Fake Tracer
+
+struct FakeTracer {
+    func generateTraceID() -> String {
+        "3f59ef6fe1fe2b12dd84ec1452696599"
+    }
+
+    struct Middleware: InstrumentationMiddlewareProtocol {
+        private let tracer: FakeTracer
+
+        init(tracer: FakeTracer) {
+            self.tracer = tracer
+        }
+
+        func extract(from headers: HTTPHeaders, into context: inout Context) {
+            let traceID = headers.first(where: { $0.0 == FakeTraceID.headerName })?.1 ?? tracer.generateTraceID()
+            context.inject(FakeTraceID.self, value: traceID)
```

Comment:
food for thought (no need to act on this now);

it may be useful that we use the words "inject" "extract" only in a single API - what currently is the Middleware (and seems like the right spot for it as well) - and the calls on context could be get/set. The reason for that is when one discusses an API and someone mentions "and i inject..." then we know it's about a middle ware, rather than the "set" action which does not have any logic, just sets a thing.

I think we've discussed that before when we looked at some other context library or am I misremembering that? Not urgent nor blocking this PR but food for thought :)

Reply:
This is also tracked in #12.

Reply:
Ah right thx